ruby/ext/google/protobuf_c/encode_decode.c

```diff
@@ -622,6 +622,24 @@ static const upb_pbdecodermethod *msgdef_decodermethod(Descriptor* desc) {
   return desc->fill_method;
 }
 
+// Closure of information required within an encode/decode-error callback: the
+// environment to free, and the error message template to use when raising a
+// Ruby exception.
+typedef struct {
+  upb_env* env;
+  const char* msg;
+} env_errdata;
+
+// Callback invoked by upb if any error occurs during parsing or serialization.
+static bool env_error_func(void* ud, const upb_status* status) {
+  env_errdata* errdata = ud;
+  // Free the env -- rb_raise will longjmp up the stack past the encode/decode
+  // function so it would not otherwise have been freed.
+  upb_env_uninit(errdata->env);
```

Comment:
> you don't actually see an issue with this implementation?

No, basically my conclusion at the moment is that this is a pattern that upb can and should support. All resources should be anchored in the upb_env at all times, so freeing the env should be safe even from a handler.

I hope this pans out as a long-term supportable semantic.